round-based/src/simulation/mod.rs

```diff
@@ -1,49 +1,496 @@
+                success = oks.len(),
+                failed = errs.len(),
+            );
+
+            for (i, err) in errs {
+                msg += &alloc::format!("- Party {i}: {err:?}\n");
+            }
+
+            panic!("{msg}");
+        }
+
+        SimResult(oks)
+    }
+}
+
+impl<T> SimResult<T>
+where
+    T: PartialEq + core::fmt::Debug,
+{
+    /// Checks that outputs of all parties are equally the same
+    ///
+    /// Returns the output on success (all the outputs are checked to be the same), otherwise
+    /// panics with a verbose error message.
+    ///
+    /// Panics if simulation contained zero parties.
+    pub fn expect_eq(mut self) -> T {
+        let Some(first) = self.0.first() else {
+            panic!("simulation contained zero parties");
+        };
+
+        if !self.0[1..].iter().all(|i| i == first) {
+            let mut msg = alloc::string::String::from(
+                "Simulation output didn't match expectations.\n\
+                Expected: all parties return the same output\n\
+                Actual  : some of the parties returned a different output\n\
+                Outputs :\n",
+            );
+
+            for (i, res) in self.0.iter().enumerate() {
+                msg += &alloc::format!("- Party {i}: {res:?}");
+            }
+
+            panic!("{msg}")
+        }
+
+        self.0
+            .pop()
+            .expect("we checked that the list contains at least one element")
+    }
+}
+
+impl<T> SimResult<T> {
+    /// Deconstructs the simulation result returning inner list of results
+    pub fn into_vec(self) -> Vec<T> {
+        self.0
+    }
+}
+
+impl<T> From<Vec<T>> for SimResult<T> {
+    fn from(list: Vec<T>) -> Self {
+        Self(list)
+    }
+}
+
+impl<T> From<SimResult<T>> for Vec<T> {
+    fn from(res: SimResult<T>) -> Self {
+        res.0
+    }
+}
+
+/// Simulates MPC protocol with parties defined as [state machines](crate::state_machine)
+pub struct Simulation<'a, O, M> {
+    parties: Vec<Party<'a, O, M>>,
+}
+
+enum Party<'a, O, M> {
+    Active {
+        party: Box<dyn crate::state_machine::StateMachine<Output = O, Msg = M> + 'a>,
+        wants_one_more_msg: bool,
+    },
+    Finished(O),
+}
+
+impl<'a, O, M> Simulation<'a, O, M>
+where
+    M: Clone + 'static,
+{
+    /// Creates empty simulation containing no parties
+    ///
+    /// New parties can be added via [`.add_party()`](Self::add_party)
+    pub fn empty() -> Self {
+        Self {
+            parties: Vec::new(),
+        }
+    }
+
+    /// Constructs empty simulation containing no parties, with allocated memory that can fit up to `n` parties without re-allocations
+    pub fn with_capacity(n: u16) -> Self {
+        Self {
+            parties: Vec::with_capacity(n.into()),
+        }
+    }
+
+    /// Constructs a simulation with `n` parties from async function that defines the protocol
+    ///
+    /// Each party has index `0 <= i < n` and instantiated via provided `init` function
```

Comment:
Probably should have a disclaimer that this runs with our own runtime instead of tokio